finetrainers/trainer.py

```diff
@@ -0,0 +1,542 @@
+
+        self.tokenizer = components.get("tokenizer", None)
+        self.text_encoder = components.get("text_encoder", None)
+        self.transformer = components.get("transformer", None)
+        self.vae = components.get("vae", None)
+        self.scheduler = components.get("scheduler", None)
+
+    def prepare_dataset(self) -> None:
+        logger.info("Initializing dataset and dataloader")
+
+        self.dataset = VideoDatasetWithResizing(
+            data_root=self.args.data_root,
+            caption_column=self.args.caption_column,
+            video_column=self.args.video_column,
+            resolution_buckets=self.args.video_resolution_buckets,
+            dataset_file=self.args.dataset_file,
+            id_token=self.args.id_token,
+        )
+        self.dataloader = torch.utils.data.DataLoader(
+            self.dataset,
+            batch_size=1,
+            sampler=BucketSampler(self.dataset, batch_size=self.args.batch_size, shuffle=True),
+            collate_fn=self.model_config.get("collate_fn"),
+            num_workers=self.args.dataloader_num_workers,
+            pin_memory=self.args.pin_memory,
+        )
+    
+    def prepare_trainable_parameters(self) -> None:
+        logger.info("Initializing trainable parameters")
+
+        # TODO(aryan): refactor later. for now only lora is supported
+        self.text_encoder.requires_grad_(False)
+        self.transformer.requires_grad_(False)
+        self.vae.requires_grad_(False)
+
+        # For mixed precision training we cast all non-trainable weights (vae, text_encoder and transformer) to half-precision
+        # as these weights are only used for inference, keeping weights in full precision is not required.
+        weight_dtype = torch.float32
+        if self.state.accelerator.mixed_precision == "fp16":
+            weight_dtype = torch.float16
+        elif self.state.accelerator.mixed_precision == "bf16":
+            weight_dtype = torch.bfloat16
+        
+        if torch.backends.mps.is_available() and weight_dtype == torch.bfloat16:
+            # due to pytorch#99272, MPS does not yet support bfloat16.
+            raise ValueError(
+                "Mixed precision training with bfloat16 is not supported on MPS. Please use fp16 (recommended) or fp32 instead."
+            )
+        
+        # TODO(aryan): handle torch dtype from accelerator vs model dtype
+        self.state.weight_dtype = weight_dtype
+        self.text_encoder.to(self.state.accelerator.device, dtype=weight_dtype)
+        self.transformer.to(self.state.accelerator.device, dtype=weight_dtype)
+        self.vae.to(self.state.accelerator.device, dtype=weight_dtype)
+
+        if self.args.gradient_checkpointing:
+            self.transformer.enable_gradient_checkpointing()
+        
+        transformer_lora_config = LoraConfig(
+            r=self.args.rank,
+            lora_alpha=self.args.lora_alpha,
```

Comment:
Just saying that alphas different from LoRA ranks aren't supported during diffusers-formatted LoRA loading as we don't serialize the metadata.

Reply:
Yes, but that's okay I think. The recommendation from a recent paper was to always train loras with alpha set to half the rank. In our README, we can talk about how to set the appropriate scale via attention_kwargs or set_adapter methods